nemo/collections/asr/parts/submodules/fast_rnnt_greedy_decoding.py

```diff
@@ -0,0 +1,440 @@
+# limitations under the License.
+
+from nemo.core.utils.cuda_python_utils import check_cuda_python_cuda_graphs_conditional_nodes_supported
+
+check_cuda_python_cuda_graphs_conditional_nodes_supported()
+
+import contextlib
+import ctypes
+import time
+from dataclasses import dataclass, field
+from itertools import product
+from typing import List, Optional, Tuple, Union
+
+import numpy as np
+import torch
+from cuda import cuda, cudart, nvrtc
+from omegaconf import DictConfig, OmegaConf
+
+from nemo.collections.asr.modules import rnnt_abstract
+from nemo.collections.asr.parts.utils import rnnt_utils
+from nemo.collections.asr.parts.utils.asr_confidence_utils import ConfidenceMethodConfig, ConfidenceMethodMixin
+from nemo.collections.common.parts.rnn import label_collate
+from nemo.core.classes import Typing, typecheck
+from nemo.core.neural_types import AcousticEncodedRepresentation, ElementType, HypothesisType, LengthsType, NeuralType
+from nemo.utils import logging
+
+
+def ASSERT_DRV(err):
+    if isinstance(err, cuda.CUresult):
+        if err != cuda.CUresult.CUDA_SUCCESS:
+            raise RuntimeError("Cuda Error: {}".format(err))
+    elif isinstance(err, nvrtc.nvrtcResult):
+        if err != nvrtc.nvrtcResult.NVRTC_SUCCESS:
+            raise RuntimeError("Nvrtc Error: {}".format(err))
+    else:
+        raise RuntimeError("Unknown error type: {}".format(err))
+
+
+def cu_call(f_call_out):
+    error, *others = f_call_out
+    if error != cudart.cudaError_t.cudaSuccess:
+        # import ipdb; ipdb.set_trace()
+        raise Exception(f"CUDA failure! {error}")
+    else:
+        # print("GALVEZ:", others)
+        return tuple(others)
+
+
+def run_nvrtc(kernel_string, kernel_name):
+    err, prog = nvrtc.nvrtcCreateProgram(str.encode(kernel_string), b"while_loop_conditional.cu", 0, [], [])
+
+    ASSERT_DRV(err)
+
+    # Compile program
+    # Not specifying --gpu-architecture will default us to a fairly low compute capability, which is a safe bet.
+    # Otherwise, there are ways to query the current device's compute capability.
+    # https://stackoverflow.com/questions/48283009/nvcc-get-device-compute-capability-in-runtime
+    opts = []
+    (err,) = nvrtc.nvrtcCompileProgram(prog, len(opts), opts)
+    err, size = nvrtc.nvrtcGetProgramLogSize(prog)
+    buf = b" " * size
+    (err,) = nvrtc.nvrtcGetProgramLog(prog, buf)
+    # print(buf.decode("utf-8"))
+    ASSERT_DRV(err)
+
+    # Get PTX from compilation
+    err, ptxSize = nvrtc.nvrtcGetPTXSize(prog)
+    ASSERT_DRV(err)
+    ptx = b" " * ptxSize
+    (err,) = nvrtc.nvrtcGetPTX(prog, ptx)
+    ASSERT_DRV(err)
+
+    # print("GALVEZ:PTX=")
+    # print(ptx.decode("utf-8"))
+    ptx = np.char.array(ptx)
+    err, module = cuda.cuModuleLoadData(ptx.ctypes.data)
+    ASSERT_DRV(err)
+    err, kernel = cuda.cuModuleGetFunction(module, kernel_name)
+    ASSERT_DRV(err)
+
+    return kernel
+
+
+def create_outer_for_loop_kernel():
```

Comment:
Could you provide a short comment of those kernels as to what they are doing / what are they necessary for ?

Reply:
It will help in case they need to be modified in the future for some reason